linker/src/lib.rs

```diff
@@ -1,27 +1,88 @@
+use std::{collections::BTreeSet, iter::once};
+
 use ast::{
     object::{Location, PILGraph},
-    parsed::{Expression, PILFile, PilStatement},
+    parsed::{
+        build::{direct_reference, namespaced_reference},
+        Expression, PILFile, PilStatement, SelectedExpressions,
+    },
 };
 use number::FieldElement;
 
 /// a monolithic linker which outputs a single AIR
 pub fn link<T: FieldElement>(mut graph: PILGraph<T>) -> PILFile<T> {
-    assert_eq!(graph.objects.len(), 1, "only one machine is supported");
-    let object = graph
-        .objects
-        .remove(&Location::default().join("main"))
-        .unwrap();
-    let location = Location::default().join("main");
+    let mut queue = BTreeSet::from([Location::default().join("main")]);
 
     let mut pil = vec![];
 
-    pil.push(PilStatement::Namespace(
-        0,
-        location.to_string(),
-        Expression::Number(T::from(object.degree)),
-    ));
-
-    pil.extend(object.pil);
+    while let Some(location) = queue.pop_first() {
+        let object = graph.objects.remove(&location).unwrap();
+
+        // create a namespace for this object
+        pil.push(PilStatement::Namespace(
+            0,
+            location.to_string(),
+            Expression::Number(T::from(object.degree)),
+        ));
+        pil.extend(object.pil);
+        for link in object.links {
+            // add the target submachine to the queue
+            queue.insert(location.clone().join(link.to.loc.clone()));
```

Comment:
I think every `link` above is unique, right? If yes this looks fine wrt redundant queuing.

Reply:
Yes, links are unique.